File: sdk/python/lichen/bincode.py
"""Minimal bincode encoder for Lichen transactions."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, List, Optional

from .pq import PqPublicKey, PqSignature, to_pq_signature
from .publickey import PublicKey


def _encode_u64(value: int) -> bytes:
    return value.to_bytes(8, byteorder="little", signed=False)


def _encode_u32(value: int) -> bytes:
    return value.to_bytes(4, byteorder="little", signed=False)


def _encode_u8(value: int) -> bytes:
    return bytes((value & 0xFF,))


def _encode_option_u64(value: Optional[int]) -> bytes:
    if value is None:
        return b"\x00"
    return b"\x01" + _encode_u64(value)


def _encode_bytes(data: bytes) -> bytes:
    return _encode_u64(len(data)) + data


def _encode_string(value: str) -> bytes:
    encoded = value.encode("utf-8")
    return _encode_u64(len(encoded)) + encoded


def _encode_vec(items: Iterable[bytes]) -> bytes:
    items_list = list(items)
    return _encode_u64(len(items_list)) + b"".join(items_list)


def _encode_pubkey(pubkey: PublicKey) -> bytes:
    raw = pubkey.to_bytes()
    if len(raw) != 32:
        raise ValueError("PublicKey must be 32 bytes")
    return raw


def _encode_pq_public_key(public_key: PqPublicKey) -> bytes:
    return _encode_u8(public_key.scheme_version) + _encode_bytes(public_key.bytes)


def _encode_pq_signature(signature: PqSignature) -> bytes:
    return (
        _encode_u8(signature.scheme_version)
        + _encode_pq_public_key(signature.public_key)
        + _encode_bytes(signature.sig)
    )


def _encode_hash(hex_str: str) -> bytes:
    raw = bytes.fromhex(hex_str.removeprefix("0x"))
    if len(raw) != 32:
        raise ValueError("Blockhash must be 32 bytes")
    return raw


@dataclass
class EncodedInstruction:
    program_id: PublicKey
    accounts: List[PublicKey]
    data: bytes


def encode_instruction(ix: EncodedInstruction) -> bytes:
    program_id = _encode_pubkey(ix.program_id)
    accounts = _encode_vec(_encode_pubkey(acc) for acc in ix.accounts)
    data = _encode_bytes(ix.data)
    return program_id + accounts + data


def encode_message(
    instructions: List[EncodedInstruction],
    recent_blockhash: str,
    compute_budget: Optional[int] = None,
    compute_unit_price: Optional[int] = None,
) -> bytes:
    encoded_instructions = _encode_vec(encode_instruction(ix) for ix in instructions)
    blockhash = _encode_hash(recent_blockhash)
    budget = _encode_option_u64(compute_budget)
    cu_price = _encode_option_u64(compute_unit_price)
    return encoded_instructions + blockhash + budget + cu_price


def encode_transaction(
    signatures: List[PqSignature],
    message_bytes: bytes,
    tx_type: int = 0,
) -> bytes:
    """Encode transaction matching Rust bincode format."""
    sig_bytes = [_encode_pq_signature(to_pq_signature(signature)) for signature in signatures]
    return _encode_u64(len(sig_bytes)) + b"".join(sig_bytes) + message_bytes + _encode_u32(tx_type)

File: sdk/js/src/bincode.ts
// Minimal bincode encoder for Lichen transactions

import { PublicKey } from './publickey';
import {
  bytesToHex as pqBytesToHex,
  hexToBytes as pqHexToBytes,
  PqPublicKey,
  PqSignature,
  toPqSignature,
} from './pq';
import type { Instruction, Message, Transaction } from './transaction';

const textEncoder = new TextEncoder();

function encodeU64LE(value: number | bigint): Uint8Array {
  const out = new Uint8Array(8);
  const view = new DataView(out.buffer);
  view.setBigUint64(0, BigInt(value), true);
  return out;
}

function encodeU32LE(value: number): Uint8Array {
  const out = new Uint8Array(4);
  const view = new DataView(out.buffer);
  view.setUint32(0, value, true);
  return out;
}

function encodeOptionU64(value?: number): Uint8Array {
  if (value === undefined || value === null) {
    return new Uint8Array([0x00]); // None
  }
  return concat([new Uint8Array([0x01]), encodeU64LE(value)]); // Some(value)
}

function concat(parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((sum, part) => sum + part.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const part of parts) {
    out.set(part, offset);
    offset += part.length;
  }
  return out;
}

function encodeBytes(data: Uint8Array): Uint8Array {
  return concat([encodeU64LE(data.length), data]);
}

function encodeString(value: string): Uint8Array {
  const encoded = textEncoder.encode(value);
  return concat([encodeU64LE(encoded.length), encoded]);
}

function encodeVec(items: Uint8Array[]): Uint8Array {
  return concat([encodeU64LE(items.length), ...items]);
}

export function hexToBytes(hex: string): Uint8Array {
  return pqHexToBytes(hex);
}

export function bytesToHex(bytes: Uint8Array): string {
  return pqBytesToHex(bytes);
}

function encodeU8(value: number): Uint8Array {
  return Uint8Array.of(value & 0xff);
}

function encodePubkey(pubkey: PublicKey): Uint8Array {
  const bytes = pubkey.toBytes();
  if (bytes.length !== 32) {
    throw new Error('PublicKey must be 32 bytes');
  }
  return bytes;
}

function encodeInstruction(ix: Instruction): Uint8Array {
  const programId = encodePubkey(ix.programId);
  const accounts = encodeVec(ix.accounts.map(encodePubkey));
  const data = encodeBytes(ix.data);
  return concat([programId, accounts, data]);
}

function encodePqPublicKey(publicKey: PqPublicKey): Uint8Array {
  return concat([encodeU8(publicKey.schemeVersion), encodeBytes(publicKey.toBytes())]);
}

function encodePqSignature(signature: PqSignature): Uint8Array {
  return concat([
    encodeU8(signature.schemeVersion),
    encodePqPublicKey(signature.publicKey),
    encodeBytes(signature.toBytes()),
  ]);
}

export function encodeMessage(message: Message): Uint8Array {
  const instructions = encodeVec(message.instructions.map(encodeInstruction));
  const blockhash = hexToBytes(message.recentBlockhash);
  if (blockhash.length !== 32) {
    throw new Error('Blockhash must be 32 bytes');
  }
  const computeBudget = encodeOptionU64(message.computeBudget);
  const computeUnitPrice = encodeOptionU64(message.computeUnitPrice);
  return concat([instructions, blockhash, computeBudget, computeUnitPrice]);
}

export function encodeTransaction(transaction: Transaction): Uint8Array {
  const sigBytes = transaction.signatures.map((signature) => encodePqSignature(toPqSignature(signature)));
  const encodedSigs = concat([encodeU64LE(sigBytes.length), ...sigBytes]);
  const messageBytes = encodeMessage(transaction.message);
  // tx_type: Native=0 (u32 LE)
  const txType = encodeU32LE(0);
  return concat([encodedSigs, messageBytes, txType]);
}
